### DISCUSSION TITLE
Paper size not correct

### DISCUSSION THREAD

**Discussion by agriaIT** [sumatrapdfreader/sumatrapdf#2731] [Q&A]
Hello,

I want to print a PDF via Batch/Commandline and SumatraPDF 3.4.6 to a Zebra ZT230 printer (driver 8.6.4.23832 200dpi ZPL).
If I open the PDF manually with SumatraPDF and print it, select the right paper format an size, the paper size and orientation did not work with the printer. It always print in a wrong format. The same result, if I use the command line with SumatraPDF and any parameters like noscale, shrink or fit.
In the print driver settings on Windows Server 2019 I have selected the correct paper format and size in the settings-tab and extended-tab under default-value.
If I open the PDF with Acrobat Reader or any other PDF-Viewer on the same PC, print it with the same printer and settings, select the right paper format an size I don't have any problems.
I also talked to Zebra and they mean, that it's not a problem with the driver but with the application.

I also tried other SumatraPDF-Versions but with the same result.
What can I do?

Thanks for help.

**Comment by GitHubRulesOK** [comment]
PDF is not suited to thermals especially non pdf enabled ones, however everyone considers SumatraPDF print as image a viable method !!

Ideally the pdf should be converted to text and the text sent to a thermal text printer, or not converted from text to fuzzy pdf image in the first place.
Thermal zebra does not need printTo you just (all platforms even my hand calculator,) simply copy file.zpl to port, no apps, no drivers, no fuss.

However to answer your question there are alternative windows drivers to attempt re-conversion such as Bartender Seagull that can try to correct some round trip anomalies.

SumatraPDF starts with the system default Print Form so if Letter or A4 that's where Printing starts, perfect for page language printers like A2-A6  lasers or inkjet but unsuited to Plotters or Rollfed labels.

Later versions can set Paperkind to numbered Forms so several examples how to say over-ride default to 6x4 but it is down to the Windows driver install to provide a list of formats and windows allocates the paperkind numeric. Certainly you need to consider other page related changes so it may be necessary to change the default shrink in settings.txt to none or fit for a given combination, beware the difference between `none` internally and similar `noscale` externally https://github.com/sumatrapdfreader/sumatrapdf/issues/1415

**Comment by KrzaQ** [comment]
I have good experiences with using Sumatra PDF 3.1 to print PDF files on a thermal printer (a zebra clone). I tried updating to 3.4 recently, but it no longer works from the console and prints incorrect size. It does work perfectly well when printing from the GUI 🤷‍♂️

I think it has something to do with the changes in 3.3, but I didn't bisect it.

**Comment by oxyphil-x** [comment]
I have very similar problem with a TSC TE200 label printer. I use Seagull driver and SumatraPDF 3.5.2
The printer's defaults (incl. paper size) are all set up correctly. I have a pdf file with a label of A6 size. The actual label in printer is smaller, around A7 format. I just need to auto shrink it. If I print from GUI everything goes well by default, I just click "print", select the printer and don't touch any settings. The page is successfully shrunk and fits the paper.
But if I use CLI options `-print-to "\\myprintserver\TE200" -print-settings "shrink"` the page doesn't get shrunk and prints oversized (wasting 2 labels).  Will try older version, as KrzaQ wrote.

**Reply by GitHubRulesOK** [reply]
@kfomichev 
 command line use stock paper sizes from the system if you have a custom size in system forms it has a numeric ID often called paperkind.
for example I dont have a ZPL labeler but do have a driver installed
```
PrinterInformation --ListPapers "ZDesigner ZD621R-203dpi ZPL"
Paper definitions for the printer 'ZDesigner ZD621R-203dpi ZPL' on 'LEZ-HP':
Paper Name (Kind) (Size in*in) (Size mm*mm)
Custom (256) (3.00x4.09) (76.20x103.89)
```
so I can target that named printer size with `......,paperkind=256` however another printer may need a different or even same number for different paper size
so for example the same 256 here is different size, so on a given machine the kind is neither fixed nor certain to be there
```
PrinterInformation --ListPapers "Qirui QR-588"
Paper definitions for the printer 'Qirui QR-588' on 'LEZ-HP':
Paper Name (Kind) (Size in*in) (Size mm*mm)
USER (256) (6.00x4.10) (152.40x104.14)
2 x 4 (257) (4.00x2.10) (101.60x53.34)
4 x 4 (258) (4.00x4.10) (101.60x104.14)
4 x 6 (259) (6.00x4.10) (152.40x104.14)
```

**Comment by oxyphil-x** [comment]
```
USER (256) (6.00x4.10) (152.40x104.14)
2 x 4 (257) (4.00x2.10) (101.60x53.34)
4 x 4 (258) (4.00x4.10) (101.60x104.14)
4 x 6 (259) (6.00x4.10) (152.40x104.14)
```
I can't use "printerinformation", I guess it's something *nix related, but I'm on Windows.
But in my Seagull driver settings I can see similar list of 4 paper sizes, and the "user" paper size (with my correct size numbers) is selected as default. 
The main question that remains totally unclear for me is: why does the app print perfectly from GUI but doesn't in CLI ? If this had something to do with driver incompatibility, it would print incorrectly regardless of GUI/CLI, wouldn't it?

**Comment by oxyphil-x** [comment]
I can experiment with `paperkind=256` . Should I place it inside `-print-settings` ?

**Reply by GitHubRulesOK** [reply]
CLI printing is entirely down to windows and its setting, there is no negotiation like in a GUI. so System default paper is whatever the default printer is using it may be Letter it may be A4 it may be anything else. Printer info was a utility written by the developer working on paperkind and source is in GitHub but a binary was posted to drop box and link should be in  that thread https://github.com/sumatrapdfreader/sumatrapdf/issues/1812

So SumatraPDF can ask in printer settings for A4 or A3 or any other form size to be respected IF POSSIBLE.
Traditionally you only had A5 A4 A3 but with the newest versions you can apply with system names A6 and A2 to either `paper=<page size>` or more recently `paperkind=blah` where normally the blah is the form number (1-unknown) lower numbers are often 
A4 (9) (11.69x8.27) (296.93x210.06)
A7 (166) (4.12x2.91) (104.65x73.91)
A10 (169) (1.46x1.01) (37.08x25.65)
PostScript Custom Page Size (32767) (11.69x8.27) (296.93x210.06)
thus some oddly overlap

the only effect SumatraPDF setting has on the command line is the **internal setting for printing scale** thus rather than default of shrink I would alter advanced settings to be **`none`** (see comment above) then you can experiment with `noscale,` `fit,` and `shrink,` combined with the orientation `portrait,` and `landscape,` and finally the page size values as textual or numeric it may require reversing the order of trial since each of those can impact the other 2 and thus there is a new one rather than portrait or landscape (not what you think they are) is `disable-auto-rotation` since the portrait and landscape are to force a 90 degree override of autorotate.

**Reply by KrzaQ** [reply]
Something definitely changed between 3.1 and 3.4, because I used it in production for years (I think), and it broke when I moved to a new server. Once I downgraded to a known good version - 3.1 - it started working again.

**Reply by oxyphil-x** [reply]
@GitHubRulesOK ,
Thank you very much for hint.
I finally got it to work correctly with this commad line: `-print-to "\\printserver\tsc te200" -print-settings "shrink,paperkind=256"` without touching anything else. I even got an A4 page to print on a single 75x120 mm label :).  I think, the undocumented `paperkind=` option should be mentioned among other setings on CLI documentation page https://www.sumatrapdfreader.org/docs/Command-line-arguments
The issue is solved.
Though it still appears like duct tape to me. I might have misunderstood your previous posts (English is not native for me), but  as I understand, this behavior is by design: the working mechanism which successfully prints in GUI mode through some Windows driver interaction, refuses to do the same in CLI mode, because this Windows interaction is __disabled by intention__. Why should somebody disable it? That's the question for another thread, maybe...

**Reply by oxyphil-x** [reply]
Just another 5 cents - After googling for printer settings I found a Powershell to get printer current default papersize width and height settings in human readable form (it might be relying upon WMI)
```
(Get-PrintConfiguration -printername "\\center-pc\tsc te200").PrintTicketXML
...
<psf:ScoredProperty name="psk:MediaSizeWidth">
                                <psf:Value xsi:type="xsd:integer">74960</psf:Value>
                        </psf:ScoredProperty>
                        <psf:ScoredProperty name="psk:MediaSizeHeight">
                                <psf:Value xsi:type="xsd:integer">120000</psf:Value>
                        </psf:ScoredProperty>
```

**Reply by GitHubRulesOK** [reply]
Windows printing function is not "disabled" from command line but is just as raw as before the days of DOS.
Print commands were exactly 100% same as `copy file.ext printer.port` and has not changed in 50 years.

As an example `write "hello world.docx"` add some text,save and then run
 `write /pt "hello world.docx" "Microsoft Print to PDF" "Microsoft Print to PDF" "Hello Port.pdf"`
you should get a pdf with  your text but you have no control over page size it will be the default page size and default font and default orientation

It is still the same perfection, half a century later, just without the thousands of combinations added since then, such as no longer only 80 characters to a line or sprocket holes to consider. the "features" are now variable global paper sizes need to be conveyed to the print driver for adjusting graphic scale etc.

so rather than think it is disabled it simply was never enabled.

**Comment by pythonlw** [comment]
I have found a solution, which is to process the size of the PDF before printing，
打印之前先处理一下PDF的尺寸，PDF打印出来就是完整的
def scale_pdf():
     # 这个方法是缩放PDF内容的尺寸，内容四周相当于有了一个边框
    from pypdf import PdfReader, PdfWriter
    from pypdf.generic import RectangleObject
    pdffile = r'D:\Order\2024010781372406154240\2024010781372406154240.pdf'
    reader = PdfReader(pdffile)
    writer = PdfWriter()
    for page_idx in range(len(reader.pages)):
        page = reader.pages[page_idx]
        width, height = page.mediabox.width, page.mediabox.height
        page.mediabox = RectangleObject((page.mediabox.left-20, page.mediabox.bottom-20, page.mediabox.right+20, page.mediabox.top+20))
        page.cropbox = RectangleObject((page.cropbox.left-20, page.cropbox.bottom-20, page.cropbox.right+20, page.cropbox.top+20))
        
        writer.add_page(page)
    out_file = pdffile.replace('.pdf', '') + '_1.pdf'
    with open(out_file, "wb") as f:
        writer.write(f)

def SumatraPDF():
    other_printer = r'C:\Users\administered\Desktop\SumatraPDF.exe'
    pdf_file = r'D:\Order\2024010781372406154240\2024010781372406154240_1.pdf'
   # fit ，适合页面大小，landscape 内容 横向90 度旋转，portrait 纵向旋转90度， 1x 打印1次, 1- 从第一页打印至最后一页
    args1 = [f"{other_printer}",  # 软件路径
             "-silent", # 静默打印
             "-print-to",
             f'{cprinter}', # 打印机名称
             "-print-settings",
             "fit,landscape,1-,paperkind=258,1x" 
             "-exit-on-print", # 打印完成自动退出
             "-disable-auto-rotation",
             f"{pdf_file1}"  # PDF文件路径
             ]
    subprocess.run(args1, encoding="utf-8", shell=True)
    print(f"\t|已发送至打印机：{cprinter}")

文档链接 https://www.sumatrapdfreader.org/docs/Command-line-arguments
